{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Biopython visualisation of genome alignments - biopython_visualisation.ipynb\n",
      "\n",
      "## Overview\n",
      "\n",
      "The purpose of this activity is \n",
      "\n",
      "1. to demonstrate Python/Biopython code for interacting with genome data and alignments\n",
      "2. to demonstrate visualisation of genome and genome comparison data using Biopython \n",
      "\n",
      "When you have completed this activity you should be able \n",
      "\n",
      "1. to use Biopython to render images of genome alignment, using output from **BLAST** and **MUMmer**"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Introduction\n",
      "\n",
      "Sometimes it is useful to generate genome, and genome comparison, visualisations programmatically, without using a GUI-based application such as **ACT**. This may be useful for a number of reasons, including the conduct and investigation of a large number of alignments, or if a number of different subsets of the alignment data are to be investigated independently.\n",
      "\n",
      "The Biopython libraries <http://www.biopython.org> have a specialised module called **GenomeDiagram**, which was designed for the generation of these kinds of images ([Pritchard *et al.* (2006)](http://bioinformatics.oxfordjournals.org/content/22/5/616.long)). In this activity, you will use Biopython to visualise genome alignments you generated in earlier activities."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Python Code\n",
      "\n",
      "The following cell contains Python code that is used in the activities below. You should feel welcome to use and adapt this code when tackling the course exercises."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# This Python code serves the following purpose:\n",
      "#\n",
      "# 1) Defines functions that give examples of rendering genome\n",
      "#    comparisons using Biopython\n",
      "\n",
      "# IMPORTS\n",
      "from IPython.display import Image\n",
      "\n",
      "from Bio import SeqIO\n",
      "from Bio.Graphics import GenomeDiagram\n",
      "from Bio.Graphics.GenomeDiagram import CrossLink\n",
      "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
      "\n",
      "from reportlab.lib import colors\n",
      "from reportlab.lib.colors import linearlyInterpolatedColor, Color\n",
      "from reportlab.lib.units import cm\n",
      "\n",
      "# FUNCTIONS\n",
      "# Draw a pairwise genome comparison\n",
      "def render_pairwise_genome_comparison(genome1, genome2, megablast=None,\n",
      "                                      crunchfile=None,\n",
      "                                      png=\"image.png\", pdf=\"image.pdf\",\n",
      "                                      low_ft_cut=50, low_link_cut=0,\n",
      "                                      min_aln_len=0):\n",
      "    \"\"\" This function takes the following arguments:\n",
      "    \n",
      "        - genome1, the location of an input genome sequence\n",
      "                   in FASTA format (used as the reference)\n",
      "        - genome2, the location of an input genome sequence\n",
      "                   in FASTA format (used as the comparator)\n",
      "        - megablast, the location of the corresponding \n",
      "                     alignment file, in megablast output format\n",
      "        - crunchfile, the location of the corresponding .crunch \n",
      "                      alignment file.\n",
      "        - png, the filename for an output PNG\n",
      "        - pdf, the filename for an output PDF\n",
      "        - low_ft_cut, percentage identity at which features are \n",
      "                      rendered grey\n",
      "        - low_link_cut, lowest percentage identity at which \n",
      "                        cross-links are rendered\n",
      "        - min_aln_len, smallest alignment that will be rendered\n",
      "                       as a cross-link\n",
      "          \n",
      "        and renders a pairwise alignment figure using \n",
      "        Biopython. The figure shows each genome sequence,\n",
      "        with a plot of GC content and GC skew, and the \n",
      "        alignment as cross-links.\n",
      "    \"\"\"\n",
      "    # Load sequence data, if possible\n",
      "    s1 = SeqIO.read(genome1, \"fasta\")\n",
      "    s2 = SeqIO.read(genome2, \"fasta\")\n",
      "\n",
      "    # Set up diagram, tracks, and feature sets for each genome\n",
      "    gd_diagram = GenomeDiagram.Diagram(\"Example Comparison\")\n",
      "    gd_genome1 = gd_diagram.new_track(1, name=s1.id, greytrack=False,\n",
      "                                      greytrack_label=True,\n",
      "                                      height=0.5, start=0, end=len(s1))\n",
      "    gd_genome1_ftset = gd_genome1.new_set()\n",
      "    gd_genome2 = gd_diagram.new_track(10, name=s2.id, greytrack=False,\n",
      "                                      greytrack_label=True,\n",
      "                                      height=0.5, start=0, end=len(s2))\n",
      "    gd_genome2_ftset = gd_genome2.new_set()\n",
      "    ftsets = {s1.id: gd_genome1_ftset, \n",
      "              s2.id: gd_genome2_ftset}  # convenience dictionary\n",
      "    \n",
      "    # Add whole genome sequence as a feature, on each track\n",
      "    gd_genome1_ftset.add_feature(SeqFeature(FeatureLocation(0, len(s1))),\n",
      "                                 sigil=\"BOX\", color=colors.oldlace, \n",
      "                                 label=\"False\")\n",
      "    gd_genome2_ftset.add_feature(SeqFeature(FeatureLocation(0, len(s2))),\n",
      "                                 sigil=\"BOX\", color=colors.oldlace, \n",
      "                                 label=\"False\")\n",
      "    \n",
      "    # Add alignment location features\n",
      "    # Input file is megaBLAST\n",
      "    if megablast is not None:\n",
      "        # Megablast features coloured on yellow and cyan scales\n",
      "        for sf1, sf2 in megablast_to_seqfeatures(megablast):\n",
      "            add_crosslinked_features(gd_diagram, sf1, sf2, ftsets, \n",
      "                                     min_aln_len, low_link_cut,\n",
      "                                     low_ft_cut,\n",
      "                                     refcolor=colors.yellow,\n",
      "                                     compcolor=colors.cyan)\n",
      "    # Input file is .crunch\n",
      "    if crunchfile is not None:\n",
      "        # .crunch features coloured on green and blue scales,\n",
      "        for sf1, sf2 in crunch_to_seqfeatures(crunchfile):\n",
      "            add_crosslinked_features(gd_diagram, sf1, sf2, ftsets, \n",
      "                                     min_aln_len, low_link_cut,\n",
      "                                     low_ft_cut,\n",
      "                                     refcolor=colors.green,\n",
      "                                     compcolor=colors.blue)\n",
      "\n",
      "    # Write image to files. We have to resize PNG on CentOS,\n",
      "    # as the FireFox browser/iPython combination doesn't\n",
      "    #\u00a0resize properly\n",
      "    draw(gd_diagram, (100 * cm, 40 * cm),\n",
      "         0, max(len(s1), len(s2)),\n",
      "         pdf, \"PDF\")\n",
      "    draw(gd_diagram, (20 * cm, 8 * cm),\n",
      "         0, max(len(s1), len(s2)),\n",
      "         png, \"PNG\")\n",
      "\n",
      "\n",
      "# Draw a three-way genome comparison\n",
      "def render_threeway_genome_comparison(genome1, genome2, genome3,\n",
      "                                      megablastlist=[], crunchfilelist=[],\n",
      "                                      png=\"image.png\", pdf=\"image.pdf\",\n",
      "                                      low_ft_cut=0, low_link_cut=0,\n",
      "                                      min_aln_len=0):\n",
      "    \"\"\" This function takes the following arguments:\n",
      "    \n",
      "        - genome1, the location of an input genome sequence\n",
      "                   in FASTA format\n",
      "        - genome2, the location of an input genome sequence\n",
      "                   in FASTA format\n",
      "        - genome3, the location of an input genome sequence\n",
      "                   in FASTA format\n",
      "        - megablastlist, the location(s) of the corresponding \n",
      "                     alignment file(s), in megablast output format\n",
      "        - crunchfilelist, the location(s) of the corresponding .crunch \n",
      "                      alignment file(s).\n",
      "        - png, the filename for an output PNG\n",
      "        - pdf, the filename for an output PDF\n",
      "        - low_ft_cut, percentage identity at which features are \n",
      "                      rendered grey\n",
      "        - low_link_cut, lowest percentage identity at which \n",
      "                        cross-links are rendered\n",
      "        - min_aln_len, smallest alignment that will be rendered\n",
      "                       as a cross-link\n",
      "          \n",
      "        and renders a pairwise alignment figure using \n",
      "        Biopython. The figure shows each genome sequence,\n",
      "        with a plot of GC content and GC skew, and the \n",
      "        alignment as cross-links.\n",
      "    \"\"\"\n",
      "    # Load sequence data, if possible\n",
      "    s1 = SeqIO.read(genome1, \"fasta\")\n",
      "    s2 = SeqIO.read(genome2, \"fasta\")\n",
      "    s3 = SeqIO.read(genome3, \"fasta\")\n",
      "\n",
      "    # Set up diagram, tracks, and feature sets for each genome\n",
      "    gd_diagram = GenomeDiagram.Diagram(\"Example Comparison\")\n",
      "    gd_genome1 = gd_diagram.new_track(1, name=s1.id, greytrack=False,\n",
      "                                      greytrack_label=True,\n",
      "                                      height=0.5, start=0, end=len(s1))\n",
      "    gd_genome1_ftset = gd_genome1.new_set()\n",
      "    gd_genome2 = gd_diagram.new_track(10, name=s2.id, greytrack=False,\n",
      "                                      greytrack_label=True,\n",
      "                                      height=0.5, start=0, end=len(s2))\n",
      "    gd_genome2_ftset = gd_genome2.new_set()\n",
      "    gd_genome3 = gd_diagram.new_track(20, name=s3.id, greytrack=False,\n",
      "                                      greytrack_label=True,\n",
      "                                      height=0.5, start=0, end=len(s3))\n",
      "    gd_genome3_ftset = gd_genome3.new_set()\n",
      "    ftsets = {s1.id: gd_genome1_ftset, \n",
      "              s2.id: gd_genome2_ftset,\n",
      "              s3.id: gd_genome3_ftset}  # convenience dictionary\n",
      "    \n",
      "    # Add whole genome sequence as a feature, on each track\n",
      "    gd_genome1_ftset.add_feature(SeqFeature(FeatureLocation(0, len(s1))),\n",
      "                                 sigil=\"BOX\", color=colors.oldlace, \n",
      "                                 label=\"False\")\n",
      "    gd_genome2_ftset.add_feature(SeqFeature(FeatureLocation(0, len(s2))),\n",
      "                                 sigil=\"BOX\", color=colors.oldlace, \n",
      "                                 label=\"False\")\n",
      "    gd_genome3_ftset.add_feature(SeqFeature(FeatureLocation(0, len(s3))),\n",
      "                                 sigil=\"BOX\", color=colors.oldlace, \n",
      "                                 label=\"False\")\n",
      "    \n",
      "    # Add alignment location features\n",
      "    # Input file is megaBLAST\n",
      "    for megablast in megablastlist:\n",
      "        for sf1, sf2 in megablast_to_seqfeatures(megablast):\n",
      "            add_crosslinked_features(gd_diagram, sf1, sf2, ftsets, \n",
      "                                     min_aln_len, low_link_cut,\n",
      "                                     low_ft_cut,\n",
      "                                     refcolor=colors.darkviolet,\n",
      "                                     compcolor=colors.darkviolet)\n",
      "    # Input file is .crunch\n",
      "    for crunchfile in crunchfilelist:\n",
      "        for sf1, sf2 in crunch_to_seqfeatures(crunchfile):\n",
      "            add_crosslinked_features(gd_diagram, sf1, sf2, ftsets, \n",
      "                                     min_aln_len, low_link_cut,\n",
      "                                     low_ft_cut,\n",
      "                                     refcolor=colors.darkviolet,\n",
      "                                     compcolor=colors.darkviolet)\n",
      "\n",
      "    # Write image to files. We have to resize PNG on CentOS,\n",
      "    # as the FireFox browser/iPython combination doesn't\n",
      "    #\u00a0resize properly\n",
      "    draw(gd_diagram, (100 * cm, 40 * cm),\n",
      "         0, max(len(s1), len(s2)),\n",
      "         pdf, \"PDF\")\n",
      "    draw(gd_diagram, (20 * cm, 8 * cm),\n",
      "         0, max(len(s1), len(s2)),\n",
      "         png, \"PNG\")\n",
      "\n",
      "# Draw a linear GenomeDiagram of specified size, and named\n",
      "# start and end points, and format\n",
      "def draw(diagram, pagesize, start, end, filename, outfmt):\n",
      "    diagram.draw(format=\"linear\", pagesize=pagesize,\n",
      "                 start=start, end=end, fragments=1)\n",
      "    diagram.write(filename, outfmt)\n",
      "\n",
      "        \n",
      "# Add crosslinked features to the passed diagram\n",
      "def add_crosslinked_features(diagram, sf1, sf2, ftsets, min_aln_len, low_link_cut,\n",
      "                             low_ft_cut, refcolor=colors.red, compcolor=colors.blue):\n",
      "    \"\"\" Add a crosslink between features sf1 and sf2 on the passed\n",
      "        diagram (so long as the minimum identity and length \n",
      "        criteria are passed)\n",
      "    \"\"\"\n",
      "    ft1 = ftsets[sf1.id].add_feature(sf1,\n",
      "                                     color=linearlyInterpolatedColor(colors.grey, \n",
      "                                                                     refcolor,\n",
      "                                                                     low_ft_cut, 100,\n",
      "                                                                     sf1.qualifiers['ident']))\n",
      "    ft2 = ftsets[sf2.id].add_feature(sf2,\n",
      "                                     color=linearlyInterpolatedColor(colors.grey,\n",
      "                                                                     compcolor,\n",
      "                                                                     low_ft_cut, 100,\n",
      "                                                                     sf1.qualifiers['ident']))    \n",
      "    aln_len = min(abs(sf1.location.end - sf1.location.start),\n",
      "                  abs(sf2.location.end - sf2.location.start))\n",
      "    if (sf1.qualifiers['ident'] > low_link_cut) and (aln_len > min_aln_len):\n",
      "        flip = False if sf1.strand == sf2.strand else True\n",
      "        # Can't use transparencies in Color on CentOS 6, \n",
      "        # as only Python 2.6 present.\n",
      "        if flip:\n",
      "            c1, c2 = Color(0, 0, 1), colors.blue\n",
      "        else:\n",
      "            c1, c2 = Color(1, 0, 0), colors.red\n",
      "        clcolor = linearlyInterpolatedColor(c1, c2, \n",
      "                                            low_link_cut, 100,\n",
      "                                            sf1.qualifiers['ident'])                    \n",
      "        diagram.cross_track_links.append(CrossLink(ft1, ft2, \n",
      "                                                    color=clcolor, \n",
      "                                                    flip=flip))\n",
      "    \n",
      "\n",
      "# Load in megaBLAST file data and return a set of linked SeqFeatures,\n",
      "# two per tuple\n",
      "def megablast_to_seqfeatures(filename):\n",
      "    \"\"\" Parses the passed megablast comparison file, returning a generator of \n",
      "        (ft1, ft2) SeqFeature tuples.\n",
      "        \n",
      "        - filename, the location of the megaBLAST file\n",
      "    \"\"\"\n",
      "    with open(filename, 'rU') as fh:\n",
      "        for line in [l.strip().split('\\t') for l in fh]:\n",
      "            # Get start and end of alignment regions\n",
      "            start1, end1, start2, end2 = \\\n",
      "                int(line[6]), int(line[7]), int(line[8]), int(line[9])\n",
      "            # Identify strands\n",
      "            strand1 = 1 if line[7] > line[6] else -1\n",
      "            strand2 = 1 if line[9] > line[8] else -1\n",
      "            # Generate SeqFeatures\n",
      "            sf1 = SeqFeature(FeatureLocation(start1, end1, strand=strand1),\n",
      "                             id=line[0], qualifiers={'ident': float(line[2]),\n",
      "                                                     'score': float(line[-1]),\n",
      "                                                     'eval': float(line[-2])})\n",
      "            sf2 = SeqFeature(FeatureLocation(start2, end2, strand=strand2),\n",
      "                             id=line[1], qualifiers={'ident': float(line[2]),\n",
      "                                                     'score': float(line[-1]),\n",
      "                                                     'eval': float(line[-2])})\n",
      "            yield (sf1, sf2)\n",
      "\n",
      "            \n",
      "# Load in .crunch file data and return a set of linked SeqFeatures,\n",
      "# two per tuple\n",
      "def crunch_to_seqfeatures(filename):\n",
      "    \"\"\" Parses the passed crunch comparison file, returning a generator of \n",
      "        (ft1, ft2) SeqFeature tuples.\n",
      "        \n",
      "        - filename, the location of the .crunch file\n",
      "    \"\"\"\n",
      "    with open(filename, 'rU') as fh:\n",
      "        for line in [l.strip().split() for l in fh if len(l.strip())]:\n",
      "            # Get start and end of alignment regions\n",
      "            start1, end1, start2, end2 = \\\n",
      "                int(line[2]), int(line[3]), int(line[5]), int(line[6])\n",
      "            # Identify strands\n",
      "            strand1 = 1 if line[3] > line[2] else -1\n",
      "            strand2 = 1 if line[6] > line[5] else -1\n",
      "            # Generate SeqFeatures\n",
      "            sf1 = SeqFeature(FeatureLocation(start1, end1, strand=strand1),\n",
      "                             id=line[4], qualifiers={'ident': float(line[1]),\n",
      "                                                     'score': float(line[0])})\n",
      "            sf2 = SeqFeature(FeatureLocation(start2, end2, strand=strand2),\n",
      "                             id=line[7], qualifiers={'ident': float(line[1]),\n",
      "                                                     'score': float(line[0])})\n",
      "            yield (sf1, sf2)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Activity"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "In the first whole genome alignment activity (part A), you used **megaBLAST** to align the *E. coli* O157:H7 Sakai chromosome (`NC_002695.fna`) with that of *E. coli* CFT 073 (`NC_004431.fna`), producing the output file `E_coli_megablast.tab`. You later visualised this alignment using the **ACT** package.\n",
      "\n",
      "You will now use Biopython to render a similar visualisation, as a static `.pdf` file, using the code above.\n",
      "\n",
      "In the example below, no cross-links are shown for alignments of less than 90%, or shorter than 3kbp in length. The files are written out to `example1.pdf` and `example1.png`. The `.png` will automatically be displayed below, if you would like to change parameters and rerun the cell."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "render_pairwise_genome_comparison(\"../data/NC_002695.fna\", \"../data/NC_004431.fna\",\n",
      "                                  megablast=\"../whole_genome_alignment/wga_output/E_coli_megablast.tab\",\n",
      "                                  low_link_cut=90, min_aln_len=3000, \n",
      "                                  pdf=\"example1.pdf\", png=\"example1.png\")\n",
      "Image(\"example1.png\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "As was the convention in **MUMmer** and **ACT** output, alignments in the same orientation in both sequences are shown with red cross-links; those where one aligned region is reversed with respect to the other have blue cross-links. Unlike those representations, the cross-links become more transparent as they approach the alignment identity limit given by `low_link_cut`.\n",
      "\n",
      "Unlike the other displays, I have chosen to represent the aligned regions directly on the genome as cyan (comparator) and yellow (reference) features. These are shaded, turning to grey as the percentage identity of the alignment falls.\n",
      "\n",
      "A `.pdf` format image was rendered in the file `example1.pdf`. You can open this and inspect it, to see how the rendered output scales when viewed zoomed-in. The generation of vector (e.g. `.pdf` or `.svg`) output, rather than rastered/bitmap output (e.g. `.png`, `.jpg`, `.tif` or `.gif`) is desirable for publication and high-quality printing."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "**ACTIVITY 1:** Render a similar image to that for the **megaBLAST** output above, using the `E_coli_nucmer.crunch` file you generated in an earlier activity, and writing the output to files `activity1.pdf` and `activity1.png`. Compare the output of the two alignments, for the same argument settings of `low_link_cut` and `min_aln_len`."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Use this cell to complete ACTIVITY 1\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "We can inspect a three-way **dc-megaBLAST** alignment between isolates of *Mycoplasma pneumoniae* (top), *E.coli* (middle), and *Mycobacterium tuberculosis*  in a similar way, using the `render_threeway_genome_comparison()` function defined above:"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "render_threeway_genome_comparison(\"../data/NC_000962.fna\", \"../data/NC_010468.fna\", \n",
      "                                  \"../data/NC_017504.fna\",\n",
      "                                  megablastlist=[\"data/NC_000962_vs_NC_010468.tab\",\n",
      "                                                 \"data/NC_017504_vs_NC_010468.tab\"], \n",
      "                                  pdf=\"example2.pdf\", png=\"example2.png\")\n",
      "Image(\"example2.png\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "**Activity 2:** Carry out a **dc-megaBLAST** comparison between the three *Pyrococcus* genomes in the `data` subdirectory:\n",
      "\n",
      "* `data/NC_000868.fna`: *Pyrococcus abyssi* GE5 chromosome\n",
      "* `data/NC_000961.fna`: *Pyrococcus horikoshii* OT3\n",
      "* `data/NC_003413.fna`: *Pyrococcus furiosus* DSM 3638\n",
      "\n",
      "Visualise the output using the function `render_threeway_genome_comparison()`. Comment on the results. Do you see similarities with the output shown in [Zivanovic *et al.* (2002)](http://dx.doi.org/10.1093/nar/30.9.1902)?\n",
      "\n",
      "*HINT:* You might find it useful to set `min_aln_len` to 1000 and inspect the `.pdf` output, rather than rely on the `.png`."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Use this cell to complete ACTIVITY 2\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}